### PR TITLE
opencv4: init at 4.0.0

### PR DESCRIPTION
**Pull request by basvandijk**
###### Motivation for this change

https://github.com/opencv/opencv/wiki/ChangeLog#version400

###### Things done

The following succeeds:

```
nix-build -E \
  'let pkgs = import ./. {}; \
   in pkgs.opencv4.override { \
       enableIpp = true; \
       enableUnfree = true; \
       enablePython = true; \
       enableGtk3 = true; \
       enableVtk = true; \
       enableFfmpeg = true; \
       enableGStreamer = true; \
       enableTesseract = true; \
       enableTbb = true; \
       enableGPhoto2 = true; \
       enableDC1394 = true; \
       enableDocs = true;\
     }' 
```

If I `enableOvis` the build fails on https://github.com/opencv/opencv/issues/12851.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/opencv/4.x.nix [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+, enableJPEG2K    ? true, jasper
+, enableEigen     ? true, eigen
+, enableOpenblas  ? true, openblas
+, enableContrib   ? true
+
+, enableCuda      ? (config.cudaSupport or false), cudatoolkit
+
+, enableUnfree    ? false
+, enableIpp       ? false
+, enablePython    ? false, pythonPackages
```
Comment (Ma27) [inline pkgs/development/libraries/opencv/4.x.nix]: Admittedly I'm not sure if that's a good idea. Python seems to be used often for such cases today and a full rebuild takes a certain amount of power and time and using Python is also one of the advertised ways of using `opencv4` according the documentation.

Also, how is this supposed to be used? I'd love to see it possible to simply add `opencv` to a python environment to start hacking (but there's no build product in `lib/pythonX.X`).

**Review by Ma27 — COMMENTED** [review]
Thanks a lot for adding this! :+1: 

On a side note: do you know how to fix the broken Darwin build? I don't know enough about Nix on darwin, but in case we don't get this fixed, I'd temporarily disable `darwin` support in `meta.platforms` :)

File: pkgs/development/libraries/opencv/4.x.nix (hunk 1 above)
Comment (basvandijk) [reply pkgs/development/libraries/opencv/4.x.nix]: @Ma27 thanks for reminding me to also have Python support. I forced-pushed a [`opencv4`](https://github.com/NixOS/nixpkgs/pull/51027/files#diff-2afbba83e1b14f9e0c304c769831a0e2R1294) attribute in `python-packages.nix`. Note that `4.0.0.nix` is a copy of `3.0.0.nix` so the way we handle Python is the same as with `opencv3`.

**Comment by basvandijk** [comment]
@GrahamcOfBorg build pythonPackages.opencv4

**Comment by basvandijk** [comment]
@Ma27 thanks for the review. I've made some attempts at fixing the darwin build but failed so far. 

So it looks like a linker issue:

```
[ 62%] Linking CXX shared library ../../lib/libopencv_videoio.dylib
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_NSData", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
  "_OBJC_CLASS_$_NSDate", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
  "_OBJC_CLASS_$_NSDictionary", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
  "_OBJC_CLASS_$_NSURL", referenced from:
      objc-class-ref in cap_avfoundation_mac.mm.o
ld: symbol(s) not found for architecture x86_64
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [modules/videoio/CMakeFiles/opencv_videoio.dir/build.make:193: lib/libopencv_videoio.4.0.0.dylib] Error 1
```

Apparently [`NSData` is part of the Foundation framework](https://developer.apple.com/documentation/foundation/nsdata). I don't see the `-framework foundation` anywhere in the opencv sources so I tried adding it to the makefile of the failing videoio module [here](https://github.com/opencv/opencv/blob/4.0.0/modules/videoio/CMakeLists.txt#L16). I also added `foundation` to the dependencies of `4.0.0.nix`. Unfortunately that didn't fix the problem.

Instead of disabling darwin support I could try just disabling the `videoio` module on darwin.

**Comment by basvandijk** [comment]
Also [reported this upstream](https://github.com/opencv/opencv/issues/13283).

**Comment by basvandijk** [comment]
@GrahamcOfBorg build opencv4

**Comment by basvandijk** [comment]
@GrahamcOfBorg build opencv4

**Comment by basvandijk** [comment]
Thanks to a great tip from @alyssais about `cf-private` my last force-push fixed the build failure on darwin (I just tested this on my Mac). Lets see if ofborg doesn't time out this time.

**Comment by basvandijk** [comment]
@GrahamcOfBorg build opencv4

Testing a final change to the patching of `lib/pkgconfig/opencv4.pc`...

**Comment by Ma27** [comment]
@basvandijk I just tested the Python support with the following expression (I checked out your branch locally):

``` nix
with import ./. { };
{
  env-py3 = python3.withPackages (ps: [ ps.opencv4 ]);
  env-py2 = python.withPackages (ps: [ ps.opencv4 ]);
}
```

The output path has a python directory (with a setup.py and a `cv2` directory with the python bindings), but they are not usable, as there's no built Python artifact that can be used automatically (they usually live in `lib/pythonX.X/site-packages`).

Unfortunately I don't know the package well enough to provide a reliable approach for fixing this.

Apart from that the expressions seems fine, if Borg is happy, this should be mergeable soon. Thanks a lot for the work! :+1:

**Comment by kalbasit** [comment]
> If I `enableOvis` the build fails on [opencv/opencv#12851](https://github.com/opencv/opencv/issues/12851).

The issue has been fixed in https://github.com/opencv/opencv_contrib/pull/1913, we should probably apply that as a patch.

**Comment by c0bw3b** [comment]
Side note: @basvandijk no need to manually call the bot, since you are a trusted user for ofBorg a build job will start _automatically_ every time you open a PR or push new commits to one of your PR.

**Comment by basvandijk** [comment]
@c0bw3b thanks for the heads up. I didn't know ofborg would rerun on forced pushes. Good to know.

@kalbasit I applied the fix for ovis.

@Ma27 I will look into the Python issue next.

**Comment by basvandijk** [comment]
The following issue incl. potential work-around is relevant regarding the Python issue: https://github.com/opencv/opencv/issues/13202

**Comment by Ma27** [comment]
Thanks a lot for the link, I only looked at the documentation :)

Does this result in building issues as well? In case this applies for OpenCV 4 as well, we could either comment the relevant parts out and wait for an upstream fix or add a warning during evaluatin (e.g. with `builtins.trace` that the Python bindings are currently broken).

**Comment by basvandijk** [comment]
> Does this result in building issues as well?
The library builds successfully but the python bindings are in the wrong place.
 
Before disabling or adding warnings I would like to try out the suggestion in the linked issue that brings back the old behaviour:

* `OPENCV_SKIP_PYTHON_LOADER=ON`
* `OPENCV_PYTHON3_INSTALL_PATH=...site/dist-packages location...`

I have some time to work on this tomorrow evening.

**Comment by Ma27** [comment]
> Before disabling or adding warnings I would like to try out the suggestion in the linked issue that brings back the old behaviour

That's even better! I mainly wanted to provide options what to do in case we don't get to fix it :)

**Comment by basvandijk** [comment]
@Ma27 I fixed the Python issue. Lets see if ofborg agrees:

@GrahamcOfBorg build pythonPackages.opencv4

**Pull request merged**